lib/logstash/codecs/json.rb

```diff
@@ -29,43 +28,57 @@ class LogStash::Codecs::JSON < LogStash::Codecs::Base
   # For nxlog users, you may to set this to "CP1252".
   config :charset, :validate => ::Encoding.name_list, :default => "UTF-8"
 
-  public
   def register
     @converter = LogStash::Util::Charset.new(@charset)
     @converter.logger = @logger
   end
 
-  public
-  def decode(data)
-    data = @converter.convert(data)
-    begin
-      decoded = LogStash::Json.load(data)
-      if decoded.is_a?(Array)
-        decoded.each {|item| yield(LogStash::Event.new(item)) }
-      elsif decoded.is_a?(Hash)
-        yield LogStash::Event.new(decoded)
-      else
-        @logger.info? && @logger.info("JSON codec received a scalar instead of an Arary or Object!", :data => data)
-        yield LogStash::Event.new("message" => data, "tags" => ["_jsonparsefailure"])
-      end
-
-    rescue LogStash::Json::ParserError => e
-      @logger.info("JSON parse failure. Falling back to plain-text", :error => e, :data => data)
-      yield LogStash::Event.new("message" => data, "tags" => ["_jsonparsefailure"])
-    rescue StandardError => e
-      # This should NEVER happen. But hubris has been the cause of many pipeline breaking things
-      # If something bad should happen we just don't want to crash logstash here.
-      @logger.warn("An unexpected error occurred parsing input to JSON",
-                   :input => data,
-                   :message => e.message,
-                   :class => e.class.name,
-                   :backtrace => e.backtrace)
-    end
-  end # def decode
+  def decode(data, &block)
+    parse(@converter.convert(data), &block)
+  end
 
-  public
   def encode(event)
     @on_event.call(event, event.to_json)
-  end # def encode
+  end
+
+  private
+
+  def from_json_parse(json, &block)
+    LogStash::Event.from_json(json).each { |event| yield event }
+  rescue LogStash::Json::ParserError => e
```

Comment:
why no `begin` block? I prefer the indentation that we get by using the begin, but I also understand it is a bit redundant.